### PR TITLE
Update uri-toolkit

### PR DESCRIPTION
**Pull request by dmfs**
* Use XwfueParameterList, drop ImmutableEntries
* Return or take CharSequence instead of Strings whenever appropriate
* Add `Parameters` class with static fields for all parameters used by OAuth2 in a common place

### CODE REVIEW

File: src/main/java/org/dmfs/oauth2/client/grants/ImplicitGrant.java [hunk 1]
```diff
@@ -114,14 +112,15 @@ public OAuth2AccessToken accessToken(HttpRequestExecutor executor) throws IOExce
...
+        private final CharSequence mState;
 
 
-        private AuthorizedImplicitGrant(OAuth2Client client, Uri redirectUri, OAuth2Scope scope, String state)
+        private AuthorizedImplicitGrant(OAuth2Client client, Uri redirectUri, OAuth2Scope scope, CharSequence state)
         {
             mClient = client;
             mRedirectUri = redirectUri;
             mScope = scope;
+            // convert state to String, because a CharSequence may not be serializable
```
Comment (lemonboston) [inline src/main/java/org/dmfs/oauth2/client/grants/ImplicitGrant.java]: this `CharSequence` is not converted here like the comment says

Comment (dmfs) [inline src/main/java/org/dmfs/oauth2/client/grants/ImplicitGrant.java]: It's actually the comment that's wrong, not the code. I'll fix that.

File: src/test/java/org/dmfs/oauth2/client/http/entities/XWwwFormUrlEncodedEntityTest.java [hunk 2]
```diff
@@ -121,40 +99,27 @@ public void testWriteToOutputStream() throws IOException
...
         // and ('&')
+        ParameterType<CharSequence> keyWithAmpersand = new BasicParameterType<>("key&with&and", TextValueType.INSTANCE);
         assertEquals("key%26with%26and=value%26with%26and",
-                new XWwwFormUrlEncodedEntity(new ImmutableEntry("key&with&and", "value&with&and")).toString());
+                new XWwwFormUrlEncodedEntity(keyWithAmpersand.parameter("value&with&and")).toString());
     }
-
-
-    @Test
-    public void testIgnoreEntries_whenEntryIsNull_orKeyIsNullOrEmpty_orValueIsNull()
```
Comment (lemonboston) [inline src/test/java/org/dmfs/oauth2/client/http/entities/XWwwFormUrlEncodedEntityTest.java]: Are these edge cases tested somewhere else?

File: src/test/java/org/dmfs/oauth2/client/http/requests/AuthorizationCodeTokenRequestTest.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+import static org.junit.Assert.fail;
+
+
+/**
+ * @author Marten Gajda
+ */
+public class AuthorizationCodeTokenRequestTest
+{
+    @Test
+    public void TestRequestEntity() throws Exception
```
Comment (lemonboston) [inline src/test/java/org/dmfs/oauth2/client/http/requests/AuthorizationCodeTokenRequestTest.java]: Upper case method name typo.

File: src/main/java/org/dmfs/oauth2/client/BasicOAuth2AuthorizationRequest.java [hunk 4]
```diff
@@ -108,12 +94,8 @@ public URI authorizationUri(URI authorizationEndpoint)
         // TODO: refuse to return a URI without a client id.
         try
         {
-            // this is not optimal, but we need to create two URIs to make sure we don't append our query to an existing query. That's probably the easiest way
-            // for now.
-            return URI.create(String.format("%s?%s",
-                    new URI(authorizationEndpoint.getScheme(), authorizationEndpoint.getAuthority(),
-                            authorizationEndpoint.getPath(), null, null).toASCIIString(),
-                    new XWwwFormUrlEncodedEntity(mParameters).toString()));
+            return URI.create(new URI(authorizationEndpoint.getScheme(), authorizationEndpoint.getAuthority(),
```
Comment (lemonboston) [inline src/main/java/org/dmfs/oauth2/client/BasicOAuth2AuthorizationRequest.java]: I see it's still java's URI here, so not all have been updated yet?

File: src/test/java/org/dmfs/oauth2/client/http/entities/XWwwFormUrlEncodedEntityTest.java (hunk 2 above)
Comment (dmfs) [inline src/test/java/org/dmfs/oauth2/client/http/entities/XWwwFormUrlEncodedEntityTest.java]: `null` parameters are no longer allowed. This is supposed to throw a NPE when you pass `null`.
Since `null` is not a valid input we don't need to test it explicitly.

Also, `null` values and keys are handled by the parameters themselves so this is no longer subject to `XWwwFormUrlEncodedEntity`.

File: src/main/java/org/dmfs/oauth2/client/BasicOAuth2AuthorizationRequest.java (hunk 4 above)
Comment (dmfs) [inline src/main/java/org/dmfs/oauth2/client/BasicOAuth2AuthorizationRequest.java]: No, for now I've only switched occurrences of redirect uris. URIs that will be used against http-client-essentials still need to be Java's URI until http-client-essentials uses uri-toolkit too.

File: src/test/java/org/dmfs/oauth2/client/http/requests/AuthorizationCodeTokenRequestTest.java (hunk 3 above)
Comment (dmfs) [inline src/test/java/org/dmfs/oauth2/client/http/requests/AuthorizationCodeTokenRequestTest.java]: fixed

File: src/test/java/org/dmfs/oauth2/client/http/entities/XWwwFormUrlEncodedEntityTest.java (hunk 2 above)
Comment (lemonboston) [inline src/test/java/org/dmfs/oauth2/client/http/entities/XWwwFormUrlEncodedEntityTest.java]: Ok, thanks.

File: src/main/java/org/dmfs/oauth2/client/BasicOAuth2AuthorizationRequest.java (hunk 4 above)
Comment (lemonboston) [inline src/main/java/org/dmfs/oauth2/client/BasicOAuth2AuthorizationRequest.java]: Ok, thanks.
